{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Search phrase: citrobacter freundii[organism] AND 33175[BioProject]\n",
      "done!\n"
     ]
    }
   ],
   "source": [
    "#-----USE THIS SCRIPT TO ADD A LIST OF SPECIES TO THE DATABASE \n",
    "\n",
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "from time import sleep  #---this command is optional \n",
    "import numpy as np\n",
    "\n",
    "import os\n",
    "\n",
    "#----------CHANGE WORKING DIRECTORY TO YOUR WORKING FOLDER\n",
    "\n",
    "os.chdir('your\\\\directory') \n",
    "\n",
    "\n",
    "Entrez.email=\"your.email@gmail.com\" #-----ENTER YOUR EMAIL ADDRESS\n",
    "\n",
    "ft=open(\"file_for_taxonomy\", \"a\") #----TAXONOMY OUTPUT FILE, new taxonomy will be added \n",
    "\n",
    "fs=open(\"file_for_sequences.txt\", \"a\") #-----REFERENCE SEQUENCES OUTPUT FILE, new sequences will be added\n",
    "\n",
    "i=1\n",
    "e=1\n",
    "\n",
    "                \n",
    "#----SEARCH NCBI 16S RefSeq DATABASE FOR THE LISTED SPECIES\n",
    "with open(\"your\\\\directory\\\\bacterial_species.txt\", \"r\") as ifile:\n",
    "        for line in ifile:\n",
    "            search_phrase=line.replace('\\n', '').replace('\\r', '') + \"[organism] AND 33175[BioProject]\"\n",
    "            handle = Entrez.esearch(db=\"nucleotide\", term=search_phrase, retmax=10)\n",
    "            result = Entrez.read(handle) \n",
    "            handle.close()\n",
    "            if not result[\"IdList\"]:\n",
    "                print ('Did not find %s.' %(search_phrase))\n",
    "                search_phrase=line.replace('\\n', '').replace('\\r', '') + \"[organism] AND 33317[BioProject]\"\n",
    "                handle = Entrez.esearch(db=\"nucleotide\", term=search_phrase, retmax=10)\n",
    "                result = Entrez.read(handle) \n",
    "                handle.close()  \n",
    "                if not result[\"IdList\"]:\n",
    "                    print ('Did not find %s.' %(search_phrase))\n",
    "                    search_phrase=line.replace('\\n', '').replace('\\r', '') + \"[organism] AND 16S[ALL FIELDS] AND RNA[All FIELDS]\"\n",
    "                    handle = Entrez.esearch(db=\"nucleotide\", term=search_phrase, retmax=10)\n",
    "                    result = Entrez.read(handle) \n",
    "                    handle.close() \n",
    "                    if not result[\"IdList\"]:\n",
    "                        print ('Also Did not find %s. Omitting this line' %(search_phrase))\n",
    "                        sleep (0.5)\n",
    "                        continue\n",
    "                    \n",
    "                       \n",
    "            print ('Search phrase: %s' %(search_phrase))\n",
    "            fetch_handle = Entrez.efetch(db=\"nuccore\", id=result[\"IdList\"], rettype=\"acc\", retmode=\"text\")\n",
    "            acc_ids = [id.strip() for id in fetch_handle]\n",
    "            my_acc_array=np.asarray(acc_ids)\n",
    "            \n",
    "            fetch_handle.close()\n",
    "\n",
    "            seq_len = 0\n",
    "            my_seq_array = []\n",
    "            for SId in my_acc_array:\n",
    "                handle = Entrez.efetch(db=\"nucleotide\", id=SId, rettype=\"gb\", retmode=\"fasta\")\n",
    "                record = SeqIO.read(handle, \"genbank\")\n",
    "                handle.close()\n",
    "                \n",
    "#-------SELECT THE LONGEST SEQUENCE FROM ALL FOUNDS GENBANK RECORDS \n",
    "                sequence = record.seq \n",
    "                if len(sequence) > seq_len:\n",
    "                    seq_len = len(sequence)\n",
    "                    seq_max = sequence\n",
    "                    r=record\n",
    "                    sleep(0.5)\n",
    "                    continue\n",
    "                    \n",
    "                \n",
    "#------EXTRACT TAXONOMY\n",
    "                   \n",
    "            try:\n",
    "                kingdom=r.annotations[\"taxonomy\"][0]\n",
    "            except IndexError:\n",
    "                kingdom=\" \"                    \n",
    "            try:\n",
    "                phylum=r.annotations[\"taxonomy\"][1]\n",
    "            except IndexError:\n",
    "                phylum=\" \"\n",
    "            try:\n",
    "                clas=r.annotations[\"taxonomy\"][2]\n",
    "            except IndexError:\n",
    "                clas=\" \"\n",
    "            try:\n",
    "                order=r.annotations[\"taxonomy\"][3]\n",
    "            except IndexError:\n",
    "                order=\" \"\n",
    "            try:\n",
    "                family=r.annotations[\"taxonomy\"][4]\n",
    "            except IndexError:\n",
    "                family=\" \"\n",
    "            try:\n",
    "                genus=r.annotations[\"taxonomy\"][5]\n",
    "            except IndexError:\n",
    "                genus=\" \"\n",
    "            try:\n",
    "                source=r.annotations[\"organism\"]\n",
    "            except IndexError:\n",
    "                source=\" \"\n",
    "            out_file_taxonomy=SId+\"\\tk__\"+kingdom+\"; p__\"+phylum+\"; c__\"+clas+\"; o__\"+order+\"; f__\"+family+\"; g__\"+genus+\"; s__\"+source+\"\\n\"\n",
    "            ft.write(out_file_taxonomy)\n",
    "            i+=1\n",
    "            \n",
    "#--------EXTRACT REFERENCE SEQUENCES IN TEXT FORMAT\n",
    "                        \n",
    "            out_file_sequence=\">\"+SId+\"\\n\"+str(seq_max)+\"\\n\"\n",
    "            fs.write(out_file_sequence)\n",
    "            e+=1\n",
    "            sleep(5.0)\n",
    "            continue\n",
    "            \n",
    "#-------DOWNLOAD REFERENCE SEQUENCES IN FASTA FORMAT\n",
    "\n",
    "            out_file=SId+'.fasta'\n",
    "            sleep (0.5) \n",
    "            SeqIO.write(r, out_file, \"fasta\")\n",
    "                        \n",
    "ft.close()\n",
    "          \n",
    "fs.close()\n",
    "\n",
    "\n",
    "print(\"done!\")\n",
    "                    \n",
    "                    \n",
    "\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
